Documentation/General/emergentArchitectureDesign/emergentArchitectureDesign.tex

```diff
@@ -0,0 +1,180 @@
+
+\begin{figure}[h!]
+\includegraphics[scale=.4]{backendInteraction.png}
+\caption{Example interaction frontend and backend}
+\label{fig: backendInteraction}
+\end{figure}
+
+
+\par
+The backend and the frontend communicate over \Gls{HTTP} requests in the \Gls{JSON} format.
+% We might want to note Websockets at a later point
+The REST API will be written in Java using JAX-RS and Jackson.
+Jackson serializes Java objects to a JSON string based on their attributes, and vice versa: it is able to deserialize a Java object from a JSON string.
+This allows us to use JSON in the frondend, which is ideal in \Gls{JS}, but use classes and objects in Java, which greatly improves the testability of the responses.
+The backend will be tested using Unit tests written using JUnit and Mockito.
+
+\subsubsection{Music matching algorithm}\label{MatchingAlgorithm}
+The Music matching algorithm determines and compares moods for songs, rooms and users.
+Moodcat will be developed around a learning algorithm.
+This means that Moodcat will improve over time.
+Default values will be determined by examining the audio features provided by our data set.
+Then, through a system of user votes, the system will learn and improve over time.
+
+
+\section{Hardware/software mapping}
+For navigating through the user interface and posting messages in the chatroom a keyboard and mouse, or a touch screen with onscreen keyboard is required.
+We will annotate our elements with proper \Gls{ARIA} accessibilty attributes so that assistive technologies, such as screen readers, will work with our service properly.
+
+\par
+Moodcat requires an audio interface to play music.
+To interact with the audio interface from the browser, the Web Audio API\cite{WebAudioAPI} and NgAudio are used.
+The users browser should support these techniques in order to use Moodcat.
+
+\section{Persistent data management}
+Moodcat has to handle with both song data and chatroom data.
+
+\par
+For the song we store the basic properties, like artist, song title, genre and duration.
+Furthermore each song has its own valence and arousal value, which is computed and corrected by the music matching algorithm.
+The algorithm uses some of the audio features of the track for the computation of the valence and arousal values, these features are stored with the song.
+
+\par
+For the chatrooms we store basic properties, the currently playing song, the room members and a chat message.
+
+\begin{figure}[h!]
+\includegraphics[scale=.4]{erDiagram.png}
+\caption{ER diagram for Moodcat}
+\label{fig:ER diagram of the described entities}
+\end{figure}
+
+\par
+Most of the currently used data will be stored in memory, as it would be intensive to query the database for each chat message that comes in.
+However, the data will eventually be persisted in a PostgreSQL\cite{PostgreSQL} \gls{RDB}.
+One one hand this is because it would be impractical to store our entire dataset in memory.
+% On the other hand, well the other hand is busy %
```

Comment:
What were you doing?
![scared](https://d1zlh37f1ep3tj.cloudfront.net/wp/wblob/54592E651337D2/476/3448F/H_I8GVCccN_4g2SCZn9HMg/baby.jpg)